[TASK] Changed the Customer Group Form to a Ui Component

lewisvoncken opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios,
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
This PR refactors the customer group form to use UI components.

### Manual testing scenarios
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
Before changes:
![image](https://user-images.githubusercontent.com/6040343/46744275-19da2200-ccab-11e8-97cd-4591c710d78c.png)

After changes:
![image](https://user-images.githubusercontent.com/6040343/46744290-22325d00-ccab-11e8-8d36-41f2ab602bde.png)


### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

sivaschenko commented on app/code/Magento/Customer/view/adminhtml/ui_component/customer_group_form.xml:
```diff
@@ -0,0 +1,91 @@
+    </argument>
+    <settings>
+        <buttons>
+            <button class="Magento\Customer\Block\Adminhtml\Group\Edit\BackButton" name="back"/>
```
I would avoid the creation of button php files/classes where possible, see the following example on how simple buttons can be added using XML configuration only:

        <item name="buttons" xsi:type="array">
            <item name="save" xsi:type="array">
                <item name="name" xsi:type="string">save</item>
                <item name="label" xsi:type="string" translate="true">Save</item>
                <item name="class" xsi:type="string">primary</item>
                <item name="url" xsi:type="string">*/*/save</item>
            </item>
            <item name="reset" xsi:type="array">
                <item name="name" xsi:type="string">reset</item>
                <item name="label" xsi:type="string" translate="true">Reset</item>
                <item name="class" xsi:type="string">reset</item>
                <item name="url" xsi:type="string">*/*/*</item>
                <item name="sort_order" xsi:type="string">30</item>
            </item>
            <item name="back" xsi:type="array">
                <item name="name" xsi:type="string">back</item>
                <item name="label" xsi:type="string" translate="true">Back</item>
                <item name="class" xsi:type="string">back</item>
                <item name="url" xsi:type="string">*/*/index</item>
            </item>
            <item name="save_and_continue" xsi:type="array">
                <item name="label" xsi:type="string" translate="true">Save and Continue Edit</item>
                <item name="class" xsi:type="string">save</item>
                <item name="data_attribute" xsi:type="array">
                    <item name="mage-init" xsi:type="array">
                        <item name="button" xsi:type="array">
                            <item name="event" xsi:type="string">saveAndContinueEdit</item>
                        </item>
                    </item>
                </item>
            </item>
        </item>

lewisvoncken replied:
@sivaschenko this has been done but only not for the Delete button because of the confirm popup.

sivaschenko commented on app/code/Magento/Customer/view/adminhtml/layout/customer_group_edit.xml:
```diff
@@ -0,0 +1,15 @@
+ */
+-->
+<page xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:View/Layout/etc/page_configuration.xsd">
+    <update handle="styles"/>
```
I don't think the styles update is necessary here

lewisvoncken replied:
@sivaschenko this has been removed

sivaschenko commented on app/code/Magento/Customer/Controller/Adminhtml/Group.php:
```diff
@@ -5,71 +5,48 @@
-     *
      * @var \Magento\Framework\Registry
      */
     protected $_coreRegistry;
```
Please make sure to add empty lines between property declarations and avoid unnecessary empty lines (including the first and last lines of class/function bodies) as that will fail the static tests build

lewisvoncken replied:
@sivaschenko thanks for the tip

sivaschenko commented on app/code/Magento/Customer/Controller/Adminhtml/Group.php:
```diff
@@ -5,71 +5,48 @@
-        GroupInterfaceFactory $groupDataFactory,
-        \Magento\Backend\Model\View\Result\ForwardFactory $resultForwardFactory,
-        \Magento\Framework\View\Result\PageFactory $resultPageFactory
+        \Magento\Framework\Registry $coreRegistry
```
Considering that \Magento\Framework\Registry is deprecated, @vkublytskyi do you have any suggestions on how to avoid using registry in this case?

lewisvoncken replied:
@vkublytskyi do you have any advice?

sivaschenko replied:
Actually as this registry dependency was not added in this pull request, it existed in this class before, that refactoring can be postponed

sivaschenko replied:
However, to ensure backward compatibility, existing constructor dependencies should be preserved, new dependencies should be added to the end of the list as optional

sivaschenko reviewed (changes requested):
There are a couple of propositions on how to improve this contribution. Please take a look at the comments

sivaschenko commented:
@lewisvoncken thanks for your contribution! Please take a look at the code review comments. Looking forward to your updates or feedback!

sivaschenko commented on app/code/Magento/Customer/view/adminhtml/ui_component/customer_group_form.xml:
```diff
@@ -15,11 +15,27 @@
+                <class>save primary</class>
+                <label translate="true">Save Customer Group</label>
+            </button>
+            <button name="save_and_continue">
```
I don't think that configuration for save_and continue is correct.

The reflection of the array that was returned by getButtonData() method:

        return [	
            'label' => __('Save and Continue Edit'),	
            'class' => 'save',	
            'data_attribute' => [	
                'mage-init' => [	
                    'button' => ['event' => 'saveAndContinueEdit'],	
                ],	
            ],	
            'sort_order' => 80,	
        ];

Is the following XML:

    <item name="save_and_continue" xsi:type="array">
        <item name="label" xsi:type="string" translate="true">Save and Continue Edit</item>
        <item name="class" xsi:type="string">save</item>
        <item name="data_attribute" xsi:type="array">
            <item name="mage-init" xsi:type="array">
                <item name="button" xsi:type="array">
                    <item name="event" xsi:type="string">saveAndContinueEdit</item>
                </item>
            </item>
        </item>
        <item name="sort_order" xsi:type="number">80</item>
    </item>

lewisvoncken replied:
@sivaschenko for the save_and_continue there are two option.

1. the one I have implemented
2. Second one is through a class

The solution you gave can't be used for Magento 2.2 or 2.3

sivaschenko replied:
@lewisvoncken the snippet I provided is simply a translation of the array you used in Button class to the XML. I don't think the URL "*/*/save/back/1" will properly work in this case.

lewisvoncken replied:
@sivaschenko thanks but the snippet only works for 2.1 

I have tested the current implementation and it works as expected.

sivaschenko reviewed (changes requested):
Please update the save_and_continue button XML configuration

sivaschenko commented on app/code/Magento/Customer/Controller/Adminhtml/Group/Delete.php:
```diff
@@ -1,38 +1,49 @@
-                return $resultRedirect->setPath('customer/*/');
+                // init model and delete
+                $model = $this->_objectManager->create(\Magento\Customer\Model\Group::class);
+                $model->load($groupId);
```
Model load and delete methods are deprecated, please use the \Magento\Customer\Api\GroupRepositoryInterface::deleteById method

lewisvoncken replied:
done

sivaschenko commented on app/code/Magento/Customer/Controller/Adminhtml/Group/Edit.php:
```diff
@@ -1,22 +1,69 @@
-        return $this->resultForwardFactory->create()->forward('new');
+        // 1. Get ID and create model
+        $groupId = $this->getRequest()->getParam('id');
+        $model = $this->_objectManager->create(\Magento\Customer\Model\Group::class);
```
Please avoid using object manager and deprecated method of the model, use the \Magento\Customer\Api\GroupRepositoryInterface::getById to retrieve the customer group by ID.

lewisvoncken replied:
done

sivaschenko commented on app/code/Magento/Customer/Controller/Adminhtml/Group/Edit.php:
```diff
@@ -1,22 +1,69 @@
+     * @param \Magento\Framework\Registry $coreRegistry
+     * @param \Magento\Framework\View\Result\PageFactory $resultPageFactory
+     */
+    public function __construct(
```
Please keep the constructor method signature unchanged except the new required dependencies added as optional parameters to the end of the parameters list.

lewisvoncken replied:
@sivaschenko 
so if I am correct old dependencies should stay?

sivaschenko commented on app/code/Magento/Customer/Controller/Adminhtml/Group/Save.php:
```diff
@@ -1,112 +1,76 @@
-                }
-                $customerGroup->setCode(!empty($customerGroupCode) ? $customerGroupCode : null);
-                $customerGroup->setTaxClassId($taxClass);
+            $model = $this->_objectManager->create(\Magento\Customer\Model\Group::class)->load($groupId);
```
Please use the repository to retrieve the group instance

lewisvoncken replied:
done

sivaschenko commented on app/code/Magento/Customer/Controller/Adminhtml/Group/Save.php:
```diff
@@ -1,112 +1,76 @@
-                        )
-                    );
+            try {
+                $model->save();
```
Please use the repository to save the group instance

lewisvoncken replied:
done

sivaschenko reviewed (changes requested):
Thank you for the response, please see my review comments.

sidolov commented:
Hi @lewisvoncken , will you continue progress with this PR?

lewisvoncken commented:
Hi @sidolov Yeah I will

sivaschenko reviewed (changes requested):
Hi @lewisvoncken thanks for the updates. The contributions should be backward compatible, that's why I have to ask you to:
 - Keep all existing properties and parameters (mark properties as deprecated)
 - Ensure all new constructor parameters are added as optional at the end of parameters list (even is there was no constructor declared beforehand)

Also, there are code style issues:
 - Controllers should implement marker interfaces - HttpGetActionInterface (or Post) Please return them
 - Please ensure there are no double blank lines and there is a blank line between parameters

lewisvoncken commented:
@sivaschenko 
Can you please do a new review?

lewisvoncken commented:
@magento-engcom-team give me test instance

magento-engcom-team commented:
Hi @lewisvoncken. Thank you for your request. I'm working on Magento instance for you

magento-engcom-team commented:
Hi @lewisvoncken, here is your [new Magento instance](https://pr-18516.instances.magento-community.engineering).
Admin access: [https://pr-18516.instances.magento-community.engineering/admin](https://pr-18516.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

sivaschenko commented on app/code/Magento/Customer/Block/Adminhtml/Group/Edit.php:
```diff
@@ -1,103 +0,0 @@
-/**
- * Customer group edit block
- */
-class Edit extends \Magento\Backend\Block\Widget\Form\Container
```
Please keep this block and mark as deprecated

lewisvoncken replied:
done

sivaschenko commented on app/code/Magento/Customer/Block/Adminhtml/Group/Edit/Form.php:
```diff
@@ -1,148 +0,0 @@
-/**
- * Adminhtml customer groups edit form
- */
-class Form extends \Magento\Backend\Block\Widget\Form\Generic
```
Please keep this block and mark as deprecated

lewisvoncken replied:
done

sivaschenko commented on app/code/Magento/Customer/Controller/Adminhtml/Group/NewAction.php:
```diff
@@ -1,58 +1,28 @@
      *
-     * @return int
-     */
-    protected function _initGroup()
```
please keep this function and mark as deprecated

lewisvoncken replied:
done

sivaschenko commented on app/code/Magento/Customer/Controller/Adminhtml/Group/Save.php:
```diff
@@ -35,9 +47,9 @@ public function __construct(
+        \Magento\Framework\Reflection\DataObjectProcessor $dataObjectProcessor,
+        DataPersistorInterface $dataPersistor = null
     ) {
-        $this->dataObjectProcessor = $dataObjectProcessor;
```
please keep the property assignment

lewisvoncken replied:
done

sivaschenko commented on app/code/Magento/Customer/Controller/Adminhtml/Group/Save.php:
```diff
@@ -46,67 +58,55 @@ public function __construct(
-     * @param array $customerGroupData
-     * @return void
-     */
-    protected function storeCustomerGroupDataToSession($customerGroupData)
```
please keep this function and mark as deprecated

lewisvoncken replied:
done

sivaschenko reviewed (changes requested):
Hi @lewisvoncken thanks for your changes, there are still several changes required for complete backward compatibility. I've added a comment for each of them.

magento-engcom-team commented:
Hi @sivaschenko, thank you for the review.
[ENGCOM-3152](https://jira.corp.magento.com/browse/ENGCOM-3152) has been created to process this Pull Request

lewisvoncken commented:
@sivaschenko I have just added one commit because of a missing use within a deprecated function but @magento-engcom-team besides that it still can be merged

sivaschenko commented on app/code/Magento/Customer/Controller/Adminhtml/Group/Edit.php:
```diff
@@ -1,22 +1,49 @@
+                return $resultRedirect->setPath('*/*/');
+            }
+        }
+        $this->_coreRegistry->register('customer_group', $group);
```
`$group` will not be defined if `$groupId` is not set. I.e. for new customer group

lewisvoncken replied:
thanks I will take a look at it

lewisvoncken replied:
@sivaschenko the requested change has been made.

sivaschenko replied:
@lewisvoncken thanks for the change, can you please also check the travis build fail

lewisvoncken replied:
@sivaschenko is this blocking for the first version?

sivaschenko reviewed (changes requested):
Hi @lewisvoncken there is a functional issue, please see my comment.

sivaschenko commented:
Hi @lewisvoncken , thanks for the update. There is "no such entity" exception on attempt to save customer group

sidolov commented:
Hi @lewisvoncken , I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

lewisvoncken commented:
@magento-engcom-team give me test instance

magento-engcom-team commented:
Hi @lewisvoncken. Thank you for your request. I'm working on Magento instance for you

magento-engcom-team commented:
Hi @lewisvoncken, here is your [new Magento instance](https://pr-18516.instances.magento-community.engineering).
Admin access: [https://pr-18516.instances.magento-community.engineering/admin](https://pr-18516.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

lewisvoncken committed:
[TASK] Changed the Customer Group Form to a Ui Component

lewisvoncken commented:
@magento-engcom-team give me 2.3-develop instance

magento-engcom-team commented:
Hi @lewisvoncken. Thank you for your request. I'm working on Magento 2.3-develop instance for you

magento-engcom-team commented:
Hi @lewisvoncken, here is your [Magento instance](https://i-18516-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-18516-2-3-develop.instances.magento-community.engineering/admin](https://i-18516-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

lewisvoncken committed:
[BUGFIX] Solved "no such entity" exception on attempt to save new customer group

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/not_signed)](https://cla.dev.magento.com/magento/magento2?pullRequest=18516) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=18516) before we can accept your contribution.<br/>**5** out of **9** committers have signed the CLA.<br/><br/>:white_check_mark: justin-at-bounteous<br/>:white_check_mark: vovsky<br/>:white_check_mark: ravi-chandra3197<br/>:white_check_mark: magento-engcom-team<br/>:white_check_mark: o-iegorov<br/>:x: tomreece<br/>:x: KevinBKozan<br/>:x: naydav<br/>:x: adifucan

lewisvoncken commented:
@magento-engcom-team give me test instance

magento-engcom-team commented:
Hi @lewisvoncken. Thank you for your request. I'm working on Magento instance for you

magento-engcom-team commented:
Hi @lewisvoncken, here is your [new Magento instance](https://pr-18516.instances.magento-community.engineering).
Admin access: [https://pr-18516.instances.magento-community.engineering/admin](https://pr-18516.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

lewisvoncken commented:
> Hi @lewisvoncken , thanks for the update. There is "no such entity" exception on attempt to save customer group

@sivaschenko I have just updated the Pull Request by squashing the earlier commits to one and added one extra for the "no such entity" exception.

I have tested it in the test instance. If you have any questions let me know.

https://pr-18516.instances.magento-community.engineering/admin/admin/

sivaschenko commented:
Thanks @lewisvoncken ! Can you please merge the latest mainline for the tests to run correctly

lewisvoncken committed:
Merge branch '2.3-develop' of github.com:magento/magento2 into experius-2.3-patch-customer_group_form-uicomponent

lewisvoncken commented:
@sivaschenko I have just updated the branch. When this one is accepted I will also update the backport for 2.2

Let me know if you have any questions.

sivaschenko reviewed (changes requested):
@lewisvoncken thanks for updating the branch! Please take a look at the failed tests

sidolov commented:
@lewisvoncken , I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!
